File: embedded32-sdk-python/src/embedded32/codec.py
"""
Embedded32 SDK - J1939 Codec

Encoding and decoding logic for J1939 messages.
Mirrors the JS SDK implementation.
"""

from typing import Dict, Tuple, Union
from .types import PGN, J1939Message, CANFrame


# =============================================================================
# PGN DATABASE
# =============================================================================

PGN_DATABASE: Dict[int, Dict[str, Union[str, int]]] = {
    0x00EA00: {"name": "Request", "length": 3},
    0x00EE00: {"name": "Address Claimed", "length": 8},
    0x00F004: {"name": "Electronic Engine Controller 1 (EEC1)", "length": 8},
    0x00F003: {"name": "Electronic Transmission Controller 1 (ETC1)", "length": 8},
    0x00F000: {"name": "Proprietary Transmission Status", "length": 8},
    0x00FEEE: {"name": "Engine Temperature 1 (ET1)", "length": 8},
    0x00FEF2: {"name": "Fuel Economy (FE)", "length": 8},
    0x00FECA: {"name": "DM1 - Active Diagnostic Trouble Codes", "length": 8},
    0x00FECB: {"name": "DM2 - Previously Active DTCs", "length": 8},
    0x00EF00: {"name": "Engine Control Command (Proprietary B)", "length": 8},
}


# =============================================================================
# J1939 ID PARSING
# =============================================================================

def parse_j1939_id(can_id: int) -> Dict[str, int]:
    """
    Parse a J1939 extended CAN ID.
    
    J1939 29-bit ID format:
    - Bits 28-26: Priority (3 bits)
    - Bits 25-24: Reserved/EDP/DP (2 bits)
    - Bits 23-16: PF (PDU Format, 8 bits)
    - Bits 15-8:  PS (PDU Specific, 8 bits)
    - Bits 7-0:   SA (Source Address, 8 bits)
    """
    priority = (can_id >> 26) & 0x07
    pf = (can_id >> 16) & 0xFF
    ps = (can_id >> 8) & 0xFF
    sa = can_id & 0xFF
    
    if pf < 240:
        # PDU1 format - destination specific
        pgn = pf << 8
        da = ps
        pdu1 = True
    else:
        # PDU2 format - broadcast
        pgn = (pf << 8) | ps
        da = 0xFF
        pdu1 = False
    
    return {
        "priority": priority,
        "pgn": pgn,
        "source_address": sa,
        "destination_address": da,
        "pdu1": pdu1
    }


def build_j1939_id(
    pgn: int,
    source_address: int,
    priority: int = 6,
    destination_address: int = 0xFF
) -> int:
    """Build a J1939 extended CAN ID."""
    pf = (pgn >> 8) & 0xFF
    ps = pgn & 0xFF
    
    can_id = (priority & 0x07) << 26
    can_id |= (pf << 16)
    
    if pf < 240:
        # PDU1 - use destination address as PS
        can_id |= (destination_address << 8)
    else:
        # PDU2 - use PS from PGN
        can_id |= (ps << 8)
    
    can_id |= source_address
    
    return can_id


# =============================================================================
# PGN NAME LOOKUP
# =============================================================================

def get_pgn_name(pgn: int) -> str:
    """Get PGN name from database."""
    entry = PGN_DATABASE.get(pgn)
    if entry:
        return entry["name"]
    return f"Unknown (0x{pgn:04X})"


# =============================================================================
# SPN DECODING
# =============================================================================

def decode_spns(pgn: int, data: bytes) -> Dict[str, Union[int, float, str, bool]]:
    """Decode SPNs from raw data based on PGN."""
    spns: Dict[str, Union[int, float, str, bool]] = {}
    
    if pgn == PGN.EEC1:  # 0xF004 - Electronic Engine Controller 1
        if len(data) >= 5:
            raw_speed = data[3] | (data[4] << 8)
            spns["engineSpeed"] = raw_speed * 0.125
        if len(data) >= 3:
            spns["torque"] = data[2] - 125
            
    elif pgn == PGN.ET1:  # 0xFEEE - Engine Temperature 1
        if len(data) >= 1:
            spns["coolantTemp"] = data[0] - 40
            
    elif pgn in (PGN.ETC1, 0xF000):  # Transmission
        if len(data) >= 2:
            raw_speed = data[0] | (data[1] << 8)
            spns["outputShaftSpeed"] = raw_speed * 0.125
        if len(data) >= 5:
            spns["gear"] = data[4]
            
    elif pgn == PGN.REQUEST:  # 0xEA00 - Request
        if len(data) >= 3:
            requested_pgn = data[0] | (data[1] << 8) | (data[2] << 16)
            spns["requestedPGN"] = f"0x{requested_pgn:04X}"
            
    elif pgn == PGN.ENGINE_CONTROL_CMD:  # 0xEF00 - Engine Control Command
        if len(data) >= 3:
            spns["targetRpm"] = data[0] | (data[1] << 8)
            spns["enable"] = data[2] == 1
            if len(data) >= 4 and data[3] != 0xFF:
                spns["faultFlags"] = data[3]
                spns["overheat"] = (data[3] & 0x01) == 1
            
    elif pgn == PGN.DM1:  # 0xFECA - DM1
        if len(data) >= 5:
            spns["lampStatus"] = data[0]
            spns["spn"] = data[2] | (data[3] << 8) | ((data[4] & 0xE0) << 11)
            spns["fmi"] = data[4] & 0x1F
    
    return spns


# =============================================================================
# PGN ENCODING
# =============================================================================

def encode_pgn_data(pgn: int, data: Dict[str, Union[int, float, bool]]) -> bytes:
    """Encode PGN data to raw bytes."""
    result = bytearray([0xFF] * 8)  # Default to 0xFF (not available)
    
    if pgn == PGN.REQUEST:  # 0xEA00 - Request
        req_pgn = data.get("requestedPGN", 0)
        result[0] = req_pgn & 0xFF
        result[1] = (req_pgn >> 8) & 0xFF
        result[2] = (req_pgn >> 16) & 0xFF
        return bytes(result[:3])
        
    elif pgn == PGN.ENGINE_CONTROL_CMD:  # 0xEF00 - Engine Control Command
        target_rpm = int(data.get("targetRpm", 0))
        result[0] = target_rpm & 0xFF
        result[1] = (target_rpm >> 8) & 0xFF
        result[2] = 1 if data.get("enable", False) else 0
        result[3] = int(data.get("faultFlags", 0))  # Fault injection flags
        result[4] = 0xFF  # Reserved
        result[5] = 0xFF
        result[6] = 0xFF
        result[7] = 0xFF
        
    elif pgn == PGN.EEC1:  # 0xF004 - EEC1
        rpm = data.get("engineSpeed", 0)
        scaled_rpm = int(rpm / 0.125)
        result[0] = 0xF0  # Torque mode
        result[1] = 0xFF
        result[2] = int(data.get("torque", 0)) + 125
        result[3] = scaled_rpm & 0xFF
        result[4] = (scaled_rpm >> 8) & 0xFF
        
    elif pgn == PGN.ET1:  # 0xFEEE - Engine Temperature
        result[0] = int(data.get("coolantTemp", 0)) + 40
    
    return bytes(result)


# =============================================================================
# FRAME ENCODING/DECODING
# =============================================================================

def decode_frame(frame: CANFrame) -> J1939Message:
    """Decode a CAN frame to a J1939 message."""
    import time
    
    parsed = parse_j1939_id(frame.id)
    data = frame.data if isinstance(frame.data, bytes) else bytes(frame.data)
    
    return J1939Message(
        pgn=parsed["pgn"],
        pgn_name=get_pgn_name(parsed["pgn"]),
        source_address=parsed["source_address"],
        destination_address=parsed["destination_address"],
        priority=parsed["priority"],
        spns=decode_spns(parsed["pgn"], data),
        raw=data,
        timestamp=frame.timestamp or time.time()
    )


def encode_frame(
    pgn: int,
    data: Dict[str, Union[int, float, bool]],
    source_address: int,
    priority: int = 6,
    destination_address: int = 0xFF
) -> CANFrame:
    """Encode a J1939 message to a CAN frame."""
    import time
    
    can_id = build_j1939_id(pgn, source_address, priority, destination_address)
    encoded = encode_pgn_data(pgn, data)
    
    return CANFrame(
        id=can_id,
        data=encoded,
        is_extended=True,
        timestamp=time.time()
    )

File: embedded32-sdk-js/src/j1939/codec.ts
/**
 * Embedded32 SDK - J1939 Codec
 * 
 * Encoding and decoding logic for J1939 messages.
 * Reuses definitions from @embedded32/j1939 where possible.
 */

import type { CANFrame, J1939Message, PGNData } from '../types.js';
import { PGN } from '../types.js';

/**
 * PGN Database (minimal set for SDK)
 * Extended from @embedded32/j1939 definitions
 */
const PGN_DATABASE: Record<number, { name: string; length: number }> = {
  0x00ea00: { name: "Request", length: 3 },
  0x00ee00: { name: "Address Claimed", length: 8 },
  0x00f004: { name: "Electronic Engine Controller 1 (EEC1)", length: 8 },
  0x00f003: { name: "Electronic Transmission Controller 1 (ETC1)", length: 8 },
  0x00f000: { name: "Proprietary Transmission Status", length: 8 },
  0x00feee: { name: "Engine Temperature 1 (ET1)", length: 8 },
  0x00fef2: { name: "Fuel Economy (FE)", length: 8 },
  0x00feca: { name: "DM1 - Active Diagnostic Trouble Codes", length: 8 },
  0x00fecb: { name: "DM2 - Previously Active DTCs", length: 8 },
  0x00ef00: { name: "Engine Control Command (Proprietary B)", length: 8 },
};

/**
 * Parse a J1939 extended CAN ID
 */
export function parseJ1939Id(canId: number): {
  priority: number;
  pgn: number;
  sourceAddress: number;
  destinationAddress: number;
  pdu1: boolean;
} {
  // J1939 29-bit ID format:
  // Bits 28-26: Priority (3 bits)
  // Bits 25-24: Reserved/EDP/DP (2 bits)
  // Bits 23-16: PF (PDU Format, 8 bits)
  // Bits 15-8:  PS (PDU Specific, 8 bits)
  // Bits 7-0:   SA (Source Address, 8 bits)
  
  const priority = (canId >> 26) & 0x07;
  const pf = (canId >> 16) & 0xFF;
  const ps = (canId >> 8) & 0xFF;
  const sa = canId & 0xFF;
  
  let pgn: number;
  let da: number;
  let pdu1: boolean;
  
  if (pf < 240) {
    // PDU1 format - destination specific
    pgn = pf << 8;
    da = ps;
    pdu1 = true;
  } else {
    // PDU2 format - broadcast
    pgn = (pf << 8) | ps;
    da = 0xFF;
    pdu1 = false;
  }
  
  return {
    priority,
    pgn,
    sourceAddress: sa,
    destinationAddress: da,
    pdu1
  };
}

/**
 * Build a J1939 extended CAN ID
 */
export function buildJ1939Id(
  pgn: number,
  sourceAddress: number,
  priority: number = 6,
  destinationAddress: number = 0xFF
): number {
  const pf = (pgn >> 8) & 0xFF;
  const ps = pgn & 0xFF;
  
  let canId = (priority & 0x07) << 26;
  canId |= (pf << 16);
  
  if (pf < 240) {
    // PDU1 - use destination address as PS
    canId |= (destinationAddress << 8);
  } else {
    // PDU2 - use PS from PGN
    canId |= (ps << 8);
  }
  
  canId |= sourceAddress;
  
  return canId;
}

/**
 * Get PGN name from database
 */
export function getPGNName(pgn: number): string {
  const entry = PGN_DATABASE[pgn];
  return entry?.name || `Unknown (0x${pgn.toString(16).toUpperCase()})`;
}

/**
 * Decode SPNs from raw data based on PGN
 */
export function decodeSPNs(pgn: number, data: Uint8Array): Record<string, number | string | boolean> {
  const spns: Record<string, number | string | boolean> = {};
  
  switch (pgn) {
    case PGN.EEC1: // 0xF004 - Electronic Engine Controller 1
      // Byte 3-4: Engine Speed (0.125 rpm/bit)
      if (data.length >= 5) {
        const rawSpeed = data[3] | (data[4] << 8);
        spns.engineSpeed = rawSpeed * 0.125;
      }
      // Byte 2: Driver's Demand Engine Torque (% - 125)
      if (data.length >= 3) {
        spns.torque = data[2] - 125;
      }
      break;
      
    case PGN.ET1: // 0xFEEE - Engine Temperature 1
      // Byte 0: Engine Coolant Temperature (°C - 40)
      if (data.length >= 1) {
        spns.coolantTemp = data[0] - 40;
      }
      break;
      
    case PGN.ETC1: // 0xF003 - Electronic Transmission Controller 1
    case 0xF000: // Proprietary Transmission Status
      // Byte 0-1: Transmission Output Shaft Speed
      if (data.length >= 2) {
        const rawSpeed = data[0] | (data[1] << 8);
        spns.outputShaftSpeed = rawSpeed * 0.125;
      }
      // Custom: Byte 4 = gear
      if (data.length >= 5) {
        spns.gear = data[4];
      }
      break;
      
    case PGN.REQUEST: // 0xEA00 - Request
      if (data.length >= 3) {
        const requestedPGN = data[0] | (data[1] << 8) | (data[2] << 16);
        spns.requestedPGN = `0x${requestedPGN.toString(16).toUpperCase()}`;
      }
      break;
      
    case PGN.ENGINE_CONTROL_CMD: // 0xEF00 - Engine Control Command
      if (data.length >= 3) {
        spns.targetRpm = data[0] | (data[1] << 8);
        spns.enable = data[2] === 1;
        if (data.length >= 4 && data[3] !== 0xFF) {
          spns.faultFlags = data[3];
          spns.overheat = (data[3] & 0x01) === 1;
        }
      }
      break;
      
    case PGN.DM1: // 0xFECA - DM1
      if (data.length >= 5) {
        spns.lampStatus = data[0];
        spns.spn = data[2] | (data[3] << 8) | ((data[4] & 0xE0) << 11);
        spns.fmi = data[4] & 0x1F;
      }
      break;
  }
  
  return spns;
}

/**
 * Encode PGN data to raw bytes
 */
export function encodePGNData(pgn: number, data: PGNData): Uint8Array {
  const bytes = new Uint8Array(8).fill(0xFF); // Default to 0xFF (not available)
  
  switch (pgn) {
    case PGN.REQUEST: // 0xEA00 - Request
      // Only 3 bytes needed
      const reqPgn = data.requestedPGN as number || 0;
      bytes[0] = reqPgn & 0xFF;
      bytes[1] = (reqPgn >> 8) & 0xFF;
      bytes[2] = (reqPgn >> 16) & 0xFF;
      return bytes.slice(0, 3);
      
    case PGN.ENGINE_CONTROL_CMD: // 0xEF00 - Engine Control Command
      const targetRpm = data.targetRpm || 0;
      bytes[0] = targetRpm & 0xFF;
      bytes[1] = (targetRpm >> 8) & 0xFF;
      bytes[2] = data.enable ? 1 : 0;
      bytes[3] = (data.faultFlags as number) || 0; // Fault injection flags
      bytes[4] = 0xFF; // Reserved
      bytes[5] = 0xFF;
      bytes[6] = 0xFF;
      bytes[7] = 0xFF;
      break;
      
    case PGN.EEC1: // 0xF004 - EEC1
      const rpm = data.engineSpeed || 0;
      const scaledRpm = Math.round(rpm / 0.125);
      bytes[0] = 0xF0; // Torque mode
      bytes[1] = 0xFF;
      bytes[2] = (data.torque as number || 0) + 125;
      bytes[3] = scaledRpm & 0xFF;
      bytes[4] = (scaledRpm >> 8) & 0xFF;
      break;
      
    case PGN.ET1: // 0xFEEE - Engine Temperature
      bytes[0] = (data.coolantTemp as number || 0) + 40;
      break;
  }
  
  return bytes;
}

/**
 * Decode a CAN frame to a J1939 message
 */
export function decodeFrame(frame: CANFrame): J1939Message {
  const parsed = parseJ1939Id(frame.id);
  const data = frame.data instanceof Uint8Array 
    ? frame.data 
    : new Uint8Array(frame.data);
  
  return {
    pgn: parsed.pgn,
    pgnName: getPGNName(parsed.pgn),
    sourceAddress: parsed.sourceAddress,
    destinationAddress: parsed.destinationAddress,
    priority: parsed.priority,
    spns: decodeSPNs(parsed.pgn, data),
    raw: data,
    timestamp: frame.timestamp || Date.now()
  };
}

/**
 * Encode a J1939 message to a CAN frame
 */
export function encodeFrame(
  pgn: number,
  data: PGNData,
  sourceAddress: number,
  priority: number = 6,
  destinationAddress: number = 0xFF
): CANFrame {
  const canId = buildJ1939Id(pgn, sourceAddress, priority, destinationAddress);
  const encoded = encodePGNData(pgn, data);
  
  return {
    id: canId,
    data: encoded,
    isExtended: true,
    timestamp: Date.now()
  };
}
